lib/elixir/lib/application.ex

```diff
@@ -8,78 +8,100 @@ defmodule Application do
-
-  You can learn more about the generation of application resource files in the
-  documentation of `Mix.Tasks.Compile.App`, available as well by running
-  `mix help compile.app`.
+  Developers typically interact with the application environment and its
+  callback module. Therefore those will be the topics we will cover first
+  before jumping into details about the application resource file and life-cycle.
 
   ## The application environment
 
-  The key `env` of an application resource file has a list of tuples that map
-  atoms to terms, and its contents are known as the application *environment*.
-  Note that this environment is unrelated to the operating system environment.
+  Each application has its own environment. The environment is a keyword list
+  that maps atoms to terms. Note that this environment is unrelated to the
+  operating system environment.
 
   By default, the environment of an application is an empty list. In a Mix
-  project you can set that key in `application/0`:
+  project's `mix.exs` file, you can set the `:env` key in `application/0`:
 
       def application do
         [env: [redis_host: "localhost"]]
       end
 
-  and the generated application resource file is going to have it included.
-
-  The environment is available after loading the application, which is a process
-  explained later:
+  The environment is available after loading the application. This is typically
+  done automatically for you by tools like Mix, but it can also be done explicitly
+  like so:
 
       Application.load(:APP_NAME)
       #=> :ok
 
-      Application.get_env(:APP_NAME, :redis_host)
+      Application.fetch_env!(:APP_NAME, :redis_host)
       #=> "localhost"
 
   In Mix projects, the environment of the application and its dependencies can
-  be overridden via the `config/config.exs` file. If you start the application
-  with Mix, that configuration is available at compile time, and at runtime too,
-  but take into account it is not included in the generated application resource
-  file, and it is not available if you start the application without Mix.
-
-  For example, someone using your application can override its `:redis_host`
-  environment variable as follows:
+  be overridden via the `config/config.exs` file. For example, someone using
+  your application can override its `:redis_host` environment variable as follows:
 
-      config :APP_NAME, redis_host: "redis.local"
+      import Config
+      config :APP_NAME, :redis_host, "redis.local"
 
   The function `put_env/3` allows dynamic configuration of the application
   environment, but as a rule of thumb each application is responsible for its
   own environment. Please do not use the functions in this module for directly
   accessing or modifying the environment of other applications.
 
-  The application environment can be overridden via the `-config` option of
-  `erl`, as well as command-line options, as we are going to see below.
+  ### Compile-time environment
```

Comment:
I found this paragraph a bit hazy. It took me a few passes + some skimming through the code to figure it out. I think it could be a bit more example driven, perhaps along the lines of:

1. Start by showing the sketch of `RedisClient` to demonstrate compile-time envs. 
2. Replace the current `config :APP_NAME ...` with the one that uses `System.get_env`.
3. Demonstrate how the code is not recompiled if the OS var is changed, and the app uses the old address to access redis.
4. Replace `Application.fetch_env!` with `Application.compile_env!`
5. Demonstrate how the system fails to boot when the OS var is changed.